### PR TITLE
dropping non-retryable activity/wf tasks

### PR DESCRIPTION
**Pull request by Shivs11**
## What changed?

- This PR introduces changes to drop activity + workflow tasks which are non-retryable due to certain specific errors.

## Why?

- Currently, these tasks are being added back to persistence on failure and end up wasting resources by continuous retries. 
- Manual intervention is required to terminate/cancel these workflows. 

## How did you test it?

- Added unit tests 
- Existing suite of tests

## Is hotfix candidate?

nope

### CODE REVIEW

File: service/matching/matching_engine.go [hunk 1]
```diff
@@ -703,6 +711,20 @@ func (e *matchingEngineImpl) getHistoryForQueryTask(
...
+		tag.WorkflowNamespaceID(task.event.Data.GetNamespaceId()),
+		tag.WorkflowID(task.event.Data.GetWorkflowId()),
+		tag.WorkflowRunID(task.event.Data.GetRunId()),
+		tag.WorkflowTaskQueueName(taskQueueName),
+		tag.TaskID(task.event.GetTaskId()),
+		tag.WorkflowEventID(task.event.Data.GetScheduledEventId()),
+		tag.Error(err),
+	)
+	// drop the task as otherwise task would be stuck in a retry-loop since the errors from within this helper is called are non-retryable
+	task.finish(nil)
```
Comment (yycptt) [inline service/matching/matching_engine.go:725]: We need to emit a metric in this case and add an alert for it.

File: service/matching/matching_engine.go [hunk 2]
```diff
@@ -615,6 +615,14 @@ pollLoop:
...
 		if err != nil {
 			switch err.(type) {
+			case *serviceerror.Internal:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
+			case *serviceerror.DataLoss:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
+			case *serialization.DeserializationError:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
+			case *serialization.SerializationError:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
```
Comment (yycptt) [inline service/matching/matching_engine.go:625]: nit: they can be grouped under one `case`.

Comment (yycptt) [reply service/matching/matching_engine.go:625]: hmm I guess for any new error handling logic added we should be checking wrapped error as well using errors.As()

File: service/matching/matching_engine.go (hunk 1 above)
Comment (Shivs11) [reply service/matching/matching_engine.go:725]: I also think it would be beneficial for adding metrics grouped by each error type (be it internal, dataloss, etc.) This might make visibility better.
thoughts?

File: service/matching/matching_engine.go (hunk 2 above)
Comment (Shivs11) [reply service/matching/matching_engine.go:625]: A couple of things, in my opinion, if we want to inculcate using `errors.as()` while checking for errors:

1. Currently, after looking at the code, it doesn't stand out to me if we are wrapping errors or not since it seems like we return the error type directly (for example - return `internal.serviceError(..)`). However, I could be wrong since i ma just getting my feet dirty with `history`. 
2. I agree with conducting `errors.As()` checks since they seem to be more stricter since they do a DFS on the error chain. What I do think we should do though, in`PollWorkflowTasks` and in `PollActivityTasks`, is replacing the type assertion with `errors.As()`. This makes things simpler + better imo.

wdyt @yycptt @dnr @ShahabT ?

File: service/matching/matching_engine.go (hunk 1 above)
Comment (yycptt) [reply service/matching/matching_engine.go:732]: Sure I don't have strong preference either way. This will only be triggered very rarely so we can also rely on logs to figure out details as long as the runbook is clear on what to search for. 

There's also a metrics.ServiceErrorTypeTag method you can use.

File: service/matching/matching_engine.go (hunk 2 above)
Comment (yycptt) [reply service/matching/matching_engine.go]: oh yeah I forgot this is across grpc...

File: service/matching/matching_engine.go (hunk 1 above)
Comment (Shivs11) [reply service/matching/matching_engine.go:732]: sure thing - I currently have grouped them by the error types for better visibility.

File: service/matching/matching_engine.go [hunk 3]
```diff
@@ -703,6 +706,32 @@ func (e *matchingEngineImpl) getHistoryForQueryTask(
...
+		tag.WorkflowRunID(task.event.Data.GetRunId()),
+		tag.WorkflowTaskQueueName(taskQueueName),
+		tag.TaskID(task.event.GetTaskId()),
+		tag.WorkflowEventID(task.event.Data.GetScheduledEventId()),
+		tag.Error(err),
+	)
+
+	// log down the metrics
+	if common.IsInternalError(err) {
+		metrics.TaskInternalErrorCounter.With(e.metricsHandler).Record(1, metrics.NamespaceTag(namespaceID))
```
Comment (dnr) [inline service/matching/matching_engine.go:722]: NamespaceTag should take namespace name, not id. The log line should probably use namespace name too

File: service/matching/matching_engine.go [hunk 4]
```diff
@@ -615,6 +615,9 @@ pollLoop:
 		resp, err := e.recordWorkflowTaskStarted(ctx, requestClone, task)
 		if err != nil {
 			switch err.(type) {
+			case *serviceerror.Internal, *serviceerror.DataLoss,
+				*serialization.DeserializationError, *serialization.SerializationError:
```
Comment (dnr) [inline service/matching/matching_engine.go:619]: are `serialization.DeserializationError` and `serialization.SerializationError` translated across rpc boundaries? I think only serviceerrors are...

File: service/matching/matching_engine.go [hunk 5]
```diff
@@ -703,6 +706,32 @@ func (e *matchingEngineImpl) getHistoryForQueryTask(
 	return hist, resp.GetResponse().GetNextPageToken(), err
 }
 
+func (e *matchingEngineImpl) nonRetryableErrorsDropTask(task *internalTask, taskQueueName string, namespaceID string, err error) {
```
Comment (dnr) [inline service/matching/matching_engine.go:709]: why pass namespaceID if you get it from task.event.Data?

File: service/matching/matching_engine.go (hunk 1 above)
Comment (dnr) [reply service/matching/matching_engine.go:732]: we're trying to reduce the number of namespace-scoped metrics.. maybe just a plain counter and rely on logs to go deeper?  imo, I think one metric is sufficient ("internal") and include the specific error type in the logs

File: service/matching/matching_engine.go (hunk 4 above)
Comment (Shivs11) [reply service/matching/matching_engine.go:619]: Turns out they aren't. Thanks for pointing this out. Did some digging and verified that this is the case as seen inside a `UnaryHandler`: 

>  from interceptor.go
> // If a UnaryHandler returns an error, it should either be produced by the
> // status package, or be one of the context errors. Otherwise, gRPC will use
> // codes.Unknown as the status code and err.Error() as the status message of the
> // RPC.


Now, this begs the question of how we still want to catch non-service errors (such as errors close to the persistence layer) inside `matching`. Instead of checking for the exact error messages inside `matching` as returned by the rpc, I propose to conduct the check inside `history` and convert serialization based errors into service errors.

This shall require us to have serialization based errors as a new serviceerror. I think this is pretty neat since we would catch more of these errors properly. Thoughts?

cc - @dnr @yycptt

Comment (dnr) [reply service/matching/matching_engine.go:619]: Note that there are two "serviceerror" packages, the public one in api-go, and the internal one in common/serviceerror. The latter is imported as "serviceerror**s**" (the last "s" for "server").

I think the two options are:
1. Add a new serialization error type to common/serviceerror (the internal one) and use that. It would be a subtype of codes.DataLoss.
2. Just convert serialization errors to the existing (public) serviceerror.DataLoss.

(1) is more information but I'm not sure it adds anything over just serviceerror.DataLoss so I might lean to (2)

File: common/metrics/metric_defs.go [hunk 6]
```diff
@@ -1145,6 +1145,9 @@ var (
 		"schedule_action_delay",
 		WithDescription("Delay between when scheduled actions should/actually happen"),
 	)
+	NonRetryableTasks = NewCounterDef(
```
Comment (yycptt) [inline common/metrics/metric_defs.go:1148]: nit: move it to matching section L1009
```suggestion
	TaskDroppedCounter = NewCounterDef(
```

Comment (yycptt) [reply common/metrics/metric_defs.go:1148]: Actually looks like this def is not used?

File: service/history/api/get_history_util.go [hunk 7]
```diff
@@ -139,6 +141,24 @@ func GetHistory(
...
+	case *serialization.DeserializationError:
+		// log event
+		shard.GetLogger().Error("encountered de-serialization error",
+			tag.WorkflowNamespaceID(namespaceID.String()),
+			tag.WorkflowID(execution.GetWorkflowId()),
+			tag.WorkflowRunID(execution.GetRunId()),
+			tag.Error(err),
+		)
+		return nil, nil, err
+	case *serialization.SerializationError:
```
Comment (yycptt) [inline service/history/api/get_history_util.go:153]: nit: merge all 3 into one `case`. The message can just be `encountered data corruption error`, the actual error is tagged anyway.

File: service/history/api/recordactivitytaskstarted/api.go [hunk 8]
```diff
@@ -149,6 +152,12 @@ func Invoke(
 	)
 
 	if err != nil {
+		var deserializationError *serialization.DeserializationError
+		var serializationError *serialization.SerializationError
+		// convert serialization errors to be captured as serviceerrors across gRPC calls
```
Comment (yycptt) [inline service/history/api/recordactivitytaskstarted/api.go:157]: it probably make more sense to do this in a grpc interceptor then we will have it for all apis. Can be done later. Or maybe just define them as grpc errors.

Comment (yycptt) [reply service/history/api/recordactivitytaskstarted/api.go:157]: I am also ok with removing those definitions and always use DataLoss. I checked the codebase quickly and I don't see them getting used outside the persistence layer.

File: service/history/events/cache.go [hunk 9]
```diff
@@ -220,6 +221,20 @@ func (e *CacheImpl) getHistoryEventFromStore(
...
+			tag.WorkflowID(key.WorkflowID),
+			tag.WorkflowRunID(key.RunID))
+		return nil, err
+	case *serialization.SerializationError:
+		// log event
+		e.logger.Error("encountered serialization error",
+			tag.WorkflowNamespaceID(key.NamespaceID.String()),
+			tag.WorkflowID(key.WorkflowID),
+			tag.WorkflowRunID(key.RunID))
+		return nil, err
```
Comment (yycptt) [inline service/history/events/cache.go:237]: same here, merge into one case. Also need to emit CacheFailures metric.

File: service/matching/matching_engine.go [hunk 10]
```diff
@@ -703,6 +705,25 @@ func (e *matchingEngineImpl) getHistoryForQueryTask(
...
+		tag.WorkflowID(task.event.Data.GetWorkflowId()),
+		tag.WorkflowRunID(task.event.Data.GetRunId()),
+		tag.WorkflowTaskQueueName(taskQueueName),
+		tag.TaskID(task.event.GetTaskId()),
+		tag.WorkflowEventID(task.event.Data.GetScheduledEventId()),
+		tag.Error(err),
+		tag.ErrorType(err),
+	)
+
+	metrics.TaskInternalErrorCounter.With(e.metricsHandler).Record(1, metrics.ServiceErrorTypeTag(err))
```
Comment (yycptt) [inline service/matching/matching_engine.go:721]: TaskInternalErrorCounter is meant for internal error only. You meant to use the NonRetryableTasks metric defined above?

File: service/history/api/recordactivitytaskstarted/api.go (hunk 8 above)
Comment (Shivs11) [reply service/history/api/recordactivitytaskstarted/api.go:157]: I completely agree on removing these errors and just catching DataLoss errors. However, I double-checked the code base too and found out that Deserialization errors are being used outside of the persistence layer (for example -`common/nexus/payload_serializer`)

No harm in changing the error returned while un-marshalling JSON to dataloss, imo, but I wanted to double check before authoring this change. 

tldr - there are places where deserialization errors are being used across non-RPC domains. Do we still want to eradicate them completely?

File: common/metrics/metric_defs.go (hunk 6 above)
Comment (Shivs11) [reply common/metrics/metric_defs.go:1148]: sorry - I intended on emitting this def as a metric later on in my code. Changed to reflect that.

File: service/history/api/recordactivitytaskstarted/api.go (hunk 8 above)
Comment (Shivs11) [reply service/history/api/recordactivitytaskstarted/api.go:157]: I personally like the idea of catching these errors in the interceptor and then converting them to a dataloss. Shall be going ahead with that.

**Review by yycptt — APPROVED** [review]

File: service/matching/matching_engine.go [hunk 11]
```diff
@@ -615,6 +615,10 @@ pollLoop:
 		resp, err := e.recordWorkflowTaskStarted(ctx, requestClone, task)
 		if err != nil {
 			switch err.(type) {
+			case *serviceerror.Internal, *serviceerror.DataLoss:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
```
Comment (dnr) [inline service/matching/matching_engine.go:619]: I would just inline the whole thing, like the other clauses. but fine either way

File: service/matching/matching_engine_test.go [hunk 12]
```diff
@@ -687,6 +690,119 @@ func (s *matchingEngineSuite) TestPollWorkflowTaskQueues_NamespaceHandover() {
...
+	}
+
+	// add an activity task
+	_, _, err := s.matchingEngine.AddActivityTask(context.Background(), &addRequest)
+	s.NoError(err)
+	s.EqualValues(s.taskManager.getTaskCount(newUnversionedRootQueueKey(namespaceId, tl, enumspb.TASK_QUEUE_TYPE_ACTIVITY)), 1)
+
+	// task is dropped with no retry; RecordActivityTaskStarted should only be called once
+	s.mockHistoryClient.EXPECT().RecordActivityTaskStarted(gomock.Any(), gomock.Any(), gomock.Any()).
+		Return(nil, serviceerror.NewInternal("Internal error")).Times(1)
```
Comment (dnr) [inline service/matching/matching_engine_test.go:713]: just curious, does this test actually fail if this returns, say, an Unavailable error? I'm wondering how deeply we need to check the "no retry" part

**Review by dnr — APPROVED** [review]

File: service/matching/matching_engine.go (hunk 11 above)
Comment (Shivs11) [reply service/matching/matching_engine.go:619]: feel like it might get too clustered + the other cases also give that notion right now. Since we are also emitting a metric in this case, let's go ahead with keeping the func call?

File: service/matching/matching_engine_test.go (hunk 12 above)
Comment (Shivs11) [reply service/matching/matching_engine_test.go:713]: Yeah, the test fails if we return that error. Personally don't think we need to go more deeper, the only place I can think of adding more testing could be the gRPC interceptor but I feel we don't need to given that we are only checking and converting an error type. 

open to hearing your thoughts on this.

File: service/matching/matching_engine.go (hunk 11 above)
Comment (dnr) [reply service/matching/matching_engine.go:619]: some refactoring with common tags might help, but no worries

File: service/matching/matching_engine_test.go (hunk 12 above)
Comment (dnr) [reply service/matching/matching_engine_test.go:713]: I think it's fine, it was just a little subtle to think about how the test checks the right thing (e.g. is there a delay before the retry?)
